src/NuGet.Core/NuGet.Packaging/Rules/NoRefOrLibFolderInPackageRule.cs

```diff
@@ -0,0 +1,111 @@
+        {
+            var files = package.GetFiles();
+            return Validate(files);
+        }
+
+        internal IEnumerable<PackagingLogMessage> Validate(IEnumerable<string> files)
+        {
+            var managedCodeConventions = new ManagedCodeConventions(new RuntimeGraph());
+            var collection = new ContentItemCollection();
+            collection.Load(files);
+
+            var libItems = GetContentForPattern(collection, managedCodeConventions.Patterns.CompileLibAssemblies);
+            var refItems = GetContentForPattern(collection, managedCodeConventions.Patterns.CompileRefAssemblies);
+            var buildItems = GetContentForPattern(collection, managedCodeConventions.Patterns.MSBuildFiles);
+
+            var libFrameworks = GetGroupFrameworks(libItems).ToArray();
+            var refFrameworks = GetGroupFrameworks(refItems).ToArray();
+            var buildFrameworks = GetGroupFrameworks(buildItems).ToArray();
+
+            if (libFrameworks.Length == 0 && refFrameworks.Length == 0)
+            {
+                //if you can't find the ref and lib folder, then find the build folder
+                if (buildFrameworks.Length != 0)
+                {
+                    //if you can find any folders other than native or any, raise an NU5127
+                    if (buildFrameworks.Any(t => (FrameworkConstants.DotNetAll.Satisfies(t) || FrameworkNameValidatorUtility.IsValidFrameworkName(t))
+                                            && t.GetShortFolderName() != FrameworkConstants.FrameworkIdentifiers.DotNet
+                                            && t.GetShortFolderName() != FrameworkConstants.FrameworkIdentifiers.Native))
+                    {
+                        var possibleFrameworks = buildFrameworks.
+                            Where(t => t.IsSpecificFramework
+                                    && t.GetShortFolderName() != FrameworkConstants.FrameworkIdentifiers.DotNet
+                                    && t.GetShortFolderName() != FrameworkConstants.FrameworkIdentifiers.Native).
+                            Select(t => t.GetShortFolderName()).ToArray();
+
+                        (var tfmNames, var suggestedDirectories) = GenerateWarningString(possibleFrameworks);
+                        
+                        var issue = new List<PackagingLogMessage>();
+                        issue.Add(PackagingLogMessage.CreateWarning(string.Format(MessageFormat, tfmNames, suggestedDirectories),
+                            NuGetLogCode.NU5127));
+                        return issue;
+                    }
+                }
+            }
+
+            return Array.Empty<PackagingLogMessage>();
+        }
+
+        private (string, string) GenerateWarningString(string[] possibleFrameworks)
+        {
+            string tfmNames = possibleFrameworks.Length > 1
+                ? string.Join(", ", possibleFrameworks, 0, possibleFrameworks.Length)
```

Comment:
when using the whole collection, you can use the `string.Join(", ", possibleFrameworks)` overload.